File: ios/Sources/UI/Pairing/PairFlowViewModel.swift
import Foundation
import Combine
import CryptoKit

@MainActor
public final class PairFlowViewModel: ObservableObject {
    @Published public var state: PairState = .idle
    @Published public var errorMessage: String?
    @Published public var timeRemaining: Int = 10

    private let networkService: NetworkService
    private let pictogramDerivation: SessionPictogramDerivation
    private let cryptoService: CryptoService
    private let trustStorage: TrustStorageService
    private var timerTask: Task<Void, Never>?
    private var expiryDate: Date?

    public init(
        networkService: NetworkService,
        pictogramDerivation: SessionPictogramDerivation,
        cryptoService: CryptoService,
        trustStorage: TrustStorageService
    ) {
        self.networkService = networkService
        self.pictogramDerivation = pictogramDerivation
        self.cryptoService = cryptoService
        self.trustStorage = trustStorage
    }

    public func startPair(serverURL: URL) async {
        state = .loading
        errorMessage = nil

        do {
            let deviceKeyPair = try await cryptoService.getDeviceKeyPair()
            let clientPublicKey = deviceKeyPair.publicKey

            let initResponse = try await networkService.initiatePair(
                baseURL: serverURL,
                clientPublicKey: clientPublicKey
            )

            guard let serverPublicKey = Data(base64Encoded: initResponse.serverPublicKey),
                  let serverNonce = Data(base64Encoded: initResponse.serverNonce) else {
                throw NetworkError.decodingError("Invalid base64 in server response")
            }

            let derivedPictogram = try pictogramDerivation.derive(
                serverPublicKey: serverPublicKey,
                clientPublicKey: clientPublicKey,
                serverNonce: serverNonce
            )

            let serverPictogramWords = initResponse.sessionPictogram

            guard derivedPictogram.names == serverPictogramWords else {
                throw NetworkError.pictogramMismatch
            }

            expiryDate = initResponse.expiresAt
            startTimer()

            state = .awaitingConfirmation(
                pictogram: derivedPictogram,
                serverURL: serverURL,
                serverId: initResponse.serverId,
                serverPublicKey: serverPublicKey,
                serverNonce: initResponse.serverNonce,
                clientPublicKey: clientPublicKey
            )

        } catch {
            state = .error(error.localizedDescription)
            errorMessage = error.localizedDescription
        }
    }

    public func confirmPair() async {
        guard case .awaitingConfirmation(
            let pictogram,
            let serverURL,
            let serverId,
            let serverPublicKey,
            let serverNonce,
            let clientPublicKey
        ) = state else {
            return
        }

        state = .completing
        stopTimer()

        do {
            let deviceName = await UIDevice.current.name
            let osVersion = await UIDevice.current.systemVersion

            let completeRequest = PairCompleteRequest(
                serverNonce: serverNonce,
                clientPublicKey: clientPublicKey.base64EncodedString(),
                deviceInfo: .init(
                    name: String(deviceName.prefix(64)),
                    platform: "ios",
                    osVersion: osVersion
                )
            )

            // Try completion first - if admin pre-approved, succeeds immediately
            var response: PairCompleteResponse
            do {
                response = try await networkService.completePair(
                    baseURL: serverURL,
                    request: completeRequest
                )
            } catch NetworkError.pairHandshakeNotApproved {
                // Needs admin approval - start polling
                state = .awaitingAdminApproval(
                    serverURL: serverURL,
                    serverId: serverId,
                    serverPublicKey: serverPublicKey,
                    completeRequest: completeRequest,
                    pollAttempts: 0
                )

                // Poll every 2 seconds for up to 5 minutes (150 attempts)
                var attempts = 0
                let maxAttempts = 150

                while attempts < maxAttempts {
                    try await Task.sleep(for: .seconds(2))
                    attempts += 1

                    // Update state with attempt count
                    state = .awaitingAdminApproval(
                        serverURL: serverURL,
                        serverId: serverId,
                        serverPublicKey: serverPublicKey,
                        completeRequest: completeRequest,
                        pollAttempts: attempts
                    )

                    do {
                        response = try await networkService.completePair(
                            baseURL: serverURL,
                            request: completeRequest
                        )
                        // Success! Break out of poll loop
                        break
                    } catch NetworkError.pairHandshakeNotApproved {
                        // Still not approved, continue polling
                        if attempts >= maxAttempts {
                            throw NetworkError.pairHandshakeExpired
                        }
                        continue
                    } catch {
                        // Other error - fail immediately
                        throw error
                    }
                }
            }

            let serverFingerprint = Data(SHA256.hash(data: serverPublicKey))
                .map { String(format: "%02x", $0) }
                .joined()

            let trustedServer = TrustedServer(
                serverUrl: serverURL,
                serverId: serverId,
                serverPublicKey: serverPublicKey.base64EncodedString(),
                serverFingerprint: serverFingerprint,
                pairedAt: response.pairedAt
            )

            try trustStorage.saveTrustedServer(trustedServer)

            state = .paired(trustedServer)

        } catch {
            state = .error(error.localizedDescription)
            errorMessage = error.localizedDescription
        }
    }

    public func denyPair() {
        stopTimer()
        state = .denied
    }

    public func reset() {
        stopTimer()
        state = .idle
        errorMessage = nil
        timeRemaining = 10
    }

    private func startTimer() {
        stopTimer()

        timerTask = Task { @MainActor in
            while !Task.isCancelled {
                guard let expiryDate = expiryDate else { break }

                let remaining = Int(expiryDate.timeIntervalSinceNow)
                timeRemaining = max(0, remaining)

                if remaining <= 0 {
                    denyPair()
                    break
                }

                try? await Task.sleep(for: .seconds(1))
            }
        }
    }

    private func stopTimer() {
        timerTask?.cancel()
        timerTask = nil
    }

    deinit {
        stopTimer()
    }
}

public enum PairState: Equatable {
    case idle
    case loading
    case awaitingConfirmation(
        pictogram: SessionPictogram,
        serverURL: URL,
        serverId: String,
        serverPublicKey: Data,
        serverNonce: String,
        clientPublicKey: Data
    )
    case completing
    case awaitingAdminApproval(
        serverURL: URL,
        serverId: String,
        serverPublicKey: Data,
        completeRequest: PairCompleteRequest,
        pollAttempts: Int
    )
    case paired(TrustedServer)
    case denied
    case error(String)

    public static func == (lhs: PairState, rhs: PairState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle),
             (.loading, .loading),
             (.completing, .completing),
             (.denied, .denied):
            return true
        case (.awaitingConfirmation(let lhsPictogram, _, _, _, _, _),
              .awaitingConfirmation(let rhsPictogram, _, _, _, _, _)):
            return lhsPictogram == rhsPictogram
        case (.awaitingAdminApproval(_, let lhsServerId, _, _, let lhsAttempts),
              .awaitingAdminApproval(_, let rhsServerId, _, _, let rhsAttempts)):
            return lhsServerId == rhsServerId && lhsAttempts == rhsAttempts
        case (.paired(let lhsServer), .paired(let rhsServer)):
            return lhsServer == rhsServer
        case (.error(let lhsMsg), .error(let rhsMsg)):
            return lhsMsg == rhsMsg
        default:
            return false
        }
    }
}

File: android/app/src/main/kotlin/com/wagmilabs/sigil/ui/pairing/PairFlowViewModel.kt
package com.wagmilabs.sigil.ui.pairing

import android.os.Build
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.wagmilabs.sigil.core.crypto.SessionPictogramDerivation
import com.wagmilabs.sigil.core.crypto.SessionPictogram as CryptoPictogram
import com.wagmilabs.sigil.core.storage.TrustStorageService
import com.wagmilabs.sigil.network.SigilApiService
import com.wagmilabs.sigil.network.models.PairCompleteRequest
import com.wagmilabs.sigil.network.models.TrustedServer
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.security.MessageDigest
import java.util.Base64
import java.util.Date

class PairFlowViewModel(
    private val pictogramDerivation: SessionPictogramDerivation,
    private val trustStorage: TrustStorageService,
    private val getDevicePublicKey: suspend () -> ByteArray
) : ViewModel() {

    private val _state = MutableStateFlow<PairState>(PairState.Idle)
    val state: StateFlow<PairState> = _state.asStateFlow()

    private val _timeRemaining = MutableStateFlow(10)
    val timeRemaining: StateFlow<Int> = _timeRemaining.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private var timerJob: Job? = null
    private var expiryDate: Date? = null

    fun startPair(serverUrl: String) {
        viewModelScope.launch {
            _state.value = PairState.Loading
            _errorMessage.value = null

            try {
                val normalizedUrl = if (!serverUrl.startsWith("http://") && !serverUrl.startsWith("https://")) {
                    "https://$serverUrl"
                } else {
                    serverUrl
                }

                val retrofit = Retrofit.Builder()
                    .baseUrl(normalizedUrl)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build()

                val apiService = retrofit.create(SigilApiService::class.java)

                val clientPublicKey = getDevicePublicKey()
                val clientPubBase64 = Base64.getEncoder().encodeToString(clientPublicKey)

                val initResponse = apiService.initiatePair(clientPubBase64)

                if (!initResponse.isSuccessful) {
                    throw Exception("Server error: ${initResponse.code()}")
                }

                val initData = initResponse.body() ?: throw Exception("Empty response from server")

                val serverPublicKey = Base64.getDecoder().decode(initData.serverPublicKey)
                val serverNonce = Base64.getDecoder().decode(initData.serverNonce)

                val derivedPictogram = pictogramDerivation.derive(
                    serverPublicKey = serverPublicKey,
                    clientPublicKey = clientPublicKey,
                    serverNonce = serverNonce
                )

                val serverPictogramWords = initData.sessionPictogram

                if (derivedPictogram.names != serverPictogramWords) {
                    throw Exception("Pictogram mismatch - possible MITM attack")
                }

                expiryDate = initData.expiresAt
                startTimer()

                _state.value = PairState.AwaitingConfirmation(
                    pictogram = derivedPictogram,
                    serverUrl = normalizedUrl,
                    serverId = initData.serverId,
                    serverPublicKey = serverPublicKey,
                    serverNonce = initData.serverNonce,
                    clientPublicKey = clientPublicKey,
                    apiService = apiService
                )

            } catch (e: Exception) {
                _state.value = PairState.Error(e.message ?: "Unknown error")
                _errorMessage.value = e.message
            }
        }
    }

    fun confirmPair() {
        val currentState = _state.value
        if (currentState !is PairState.AwaitingConfirmation) return

        viewModelScope.launch {
            _state.value = PairState.Completing
            stopTimer()

            try {
                val deviceName = Build.MODEL.take(64)
                val osVersion = Build.VERSION.RELEASE

                val completeRequest = PairCompleteRequest(
                    serverNonce = currentState.serverNonce,
                    clientPublicKey = Base64.getEncoder().encodeToString(currentState.clientPublicKey),
                    deviceInfo = PairCompleteRequest.DeviceInfo(
                        name = deviceName,
                        platform = "android",
                        osVersion = osVersion
                    )
                )

                // Try completion first - if admin pre-approved, succeeds immediately
                var response = currentState.apiService.completePair(completeRequest)

                if (!response.isSuccessful) {
                    val errorCode = response.errorBody()?.string() // Parse error code

                    if (response.code() == 403 && errorCode?.contains("NOT_APPROVED") == true) {
                        // Needs admin approval - start polling
                        _state.value = PairState.AwaitingAdminApproval(
                            serverUrl = currentState.serverUrl,
                            serverId = currentState.serverId,
                            serverPublicKey = currentState.serverPublicKey,
                            completeRequest = completeRequest,
                            apiService = currentState.apiService,
                            pollAttempts = 0
                        )

                        // Poll every 2 seconds for up to 5 minutes (150 attempts)
                        var attempts = 0
                        val maxAttempts = 150

                        while (attempts < maxAttempts) {
                            delay(2000) // 2 seconds
                            attempts++

                            // Update state with attempt count
                            _state.value = PairState.AwaitingAdminApproval(
                                serverUrl = currentState.serverUrl,
                                serverId = currentState.serverId,
                                serverPublicKey = currentState.serverPublicKey,
                                completeRequest = completeRequest,
                                apiService = currentState.apiService,
                                pollAttempts = attempts
                            )

                            response = currentState.apiService.completePair(completeRequest)

                            if (response.isSuccessful) {
                                // Success! Break out of poll loop
                                break
                            } else if (response.code() == 403 && errorCode?.contains("NOT_APPROVED") == true) {
                                // Still not approved, continue polling
                                if (attempts >= maxAttempts) {
                                    throw Exception("Pairing timed out waiting for admin approval")
                                }
                                continue
                            } else {
                                // Other error - fail immediately
                                throw Exception("Server error: ${response.code()}")
                            }
                        }
                    } else {
                        throw Exception("Server error: ${response.code()}")
                    }
                }

                val completeData = response.body() ?: throw Exception("Empty response from server")

                val serverFingerprint = MessageDigest.getInstance("SHA-256")
                    .digest(currentState.serverPublicKey)
                    .joinToString("") { "%02x".format(it) }

                val trustedServer = TrustedServer(
                    serverUrl = currentState.serverUrl,
                    serverId = currentState.serverId,
                    serverPublicKey = Base64.getEncoder().encodeToString(currentState.serverPublicKey),
                    serverFingerprint = serverFingerprint,
                    pairedAt = completeData.pairedAt
                )

                trustStorage.saveTrustedServer(trustedServer)

                _state.value = PairState.Paired(trustedServer)

            } catch (e: Exception) {
                _state.value = PairState.Error(e.message ?: "Unknown error")
                _errorMessage.value = e.message
            }
        }
    }

    fun denyPair() {
        stopTimer()
        _state.value = PairState.Denied
    }

    fun reset() {
        stopTimer()
        _state.value = PairState.Idle
        _errorMessage.value = null
        _timeRemaining.value = 10
    }

    private fun startTimer() {
        stopTimer()

        timerJob = viewModelScope.launch {
            while (true) {
                val expiry = expiryDate ?: break
                val now = Date()
                val remaining = ((expiry.time - now.time) / 1000).toInt()
                _timeRemaining.value = maxOf(0, remaining)

                if (remaining <= 0) {
                    denyPair()
                    break
                }

                delay(1000)
            }
        }
    }

    private fun stopTimer() {
        timerJob?.cancel()
        timerJob = null
    }

    override fun onCleared() {
        super.onCleared()
        stopTimer()
    }
}

sealed class PairState {
    object Idle : PairState()
    object Loading : PairState()
    data class AwaitingConfirmation(
        val pictogram: CryptoPictogram,
        val serverUrl: String,
        val serverId: String,
        val serverPublicKey: ByteArray,
        val serverNonce: String,
        val clientPublicKey: ByteArray,
        val apiService: SigilApiService
    ) : PairState()
    object Completing : PairState()
    data class AwaitingAdminApproval(
        val serverUrl: String,
        val serverId: String,
        val serverPublicKey: ByteArray,
        val completeRequest: PairCompleteRequest,
        val apiService: SigilApiService,
        val pollAttempts: Int
    ) : PairState()
    data class Paired(val server: TrustedServer) : PairState()
    object Denied : PairState()
    data class Error(val message: String) : PairState()
}
